core/src/main/java/io/questdb/griffin/engine/functions/catalogue/TableCatalogueFunctionFactory.java

```diff
@@ -0,0 +1,249 @@
+                findFileStruct = 0;
+            }
+        }
+
+        private boolean next0() {
+            do {
+                final long pname = ff.findName(findFileStruct);
+                nativeLPSZ.of(pname);
+                if (
+                        ff.findType(findFileStruct) == Files.DT_DIR
+                                && !Chars.equals(nativeLPSZ, '.')
+                                && !Chars.equals(nativeLPSZ, "..")
+                ) {
+
+                    path.trimTo(plimit);
+                    if (ff.exists(path.concat(pname).concat(TableUtils.TXN_FILE_NAME).$())) {
+                        return true;
+                    }
+                }
+            } while (ff.findNext(findFileStruct) > 0);
+
+            ff.findClose(findFileStruct);
+            findFileStruct = 0;
+            return false;
+        }
+
+        private class TableCatalogueRecord implements Record {
+            private final StringSink utf8SinkA = new StringSink();
+            private final StringSink utf8SinkB = new StringSink();
+
+            @Override
+            public int getInt(int col) {
+                CharSequence table = getStr(0);
+                if (table != null) {
+                    TableReader reader = new TableReader(configuration, getStr(0));
+                    return reader.getPartitionCount();
+                } else {
+                    return -1;
+                }
+            }
+
+            @Override
+            public CharSequence getStr(int col) {
+                CharSequence table;
+                utf8SinkA.clear();
+                if (Chars.utf8DecodeZ(ff.findName(findFileStruct), utf8SinkA)) {
+                    table = utf8SinkA;
+                } else {
+                    return null;
+                }
+
+                if (col == 0) {
+                    return table;
+                } else {
+                    TableReader reader = new TableReader(configuration, getStr(0));
+                    return PartitionBy.toString(reader.getPartitionedBy());
+                }
+            }
+
+            @Override
+            public CharSequence getStrB(int col) {
+                CharSequence table;
+                utf8SinkB.clear();
+                if (Chars.utf8DecodeZ(ff.findName(findFileStruct), utf8SinkB)) {
+                    table = utf8SinkB;
+                } else {
+                    return null;
+                }
+
+                if (col == 0) {
+                    return table;
+                } else {
+                    TableReader reader = new TableReader(configuration, getStr(0));
+                    return PartitionBy.toString(reader.getPartitionedBy());
+                }
+            }
+
+            @Override
+            public long getLong(int col) {
+                CharSequence table = getStr(0);
+                if (table != null) {
+                    TableReader reader = new TableReader(configuration, getStr(0));
```

Comment:
It is a little excessive to create an instance of table reader on column get. The information accessed here is available from metadata, which is a small subset of what TableReader is. Perhaps alternative approach could be to read and cache metadata on cursor.next() and column getters access that.

Example of reading metadata is in TableReader class and few other places. Metadata is stored in a small file that can be memory mapped using reusable instance of ReadOnlyColumn if I remember correctly (I’m typing this from a train)

Reply:
You recall correctly. However, some metadata (like min and max timestamps) comes from the transaction file in `private boolean readTxn()`. Calculations for partition counts are spread out in a few functions across `TableReader` and probably shouldn't be copy/pasted so that results are consistent with the Java API.

I've left TableReader in, but have moved its instantiation to be in `hasNext()`.